Add mapping parameters documentation (#7115)
opensearch-project/documentation-website: _field-types/mapping-parameters/analyzer.md

---
layout: default
title: analyzer
parent: Mapping parameters
grand_parent: Mapping and field types
nav_order: 5
has_children: false
has_toc: false
---

# `analyzer`

The `analyzer` mapping parameter is used to define the text analysis process that is applied to a text field during both indexing and searching operations.

The key functions of the `analyzer` mapping parameter are:

> 1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched.

Reviewer: Caveat to call out: Each generated should not be more than 32766 bytes in length as indexing fails otherwise.

> 1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched.

Author: ```suggestion
1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched. Each generated token must not exceed 32,766 bytes to avoid indexing failures.
```

2. **Normalization:** The analyzer can apply various normalization techniques, such as converting text to lowercase, removing stopwords, and stemming/lemmatizing words.

3. **Consistency:** By defining the same analyzer for both indexing and searching, you ensure that the text analysis process is consistent, which helps improve the relevance of search results.

> 4. **Customization:** OpenSearch allows you to define custom analyzers by specifying the tokenizer, character filters, and token filters to use. This gives you fine-grained control over the text analysis process.

Reviewer: This can all lie in a different page later, which defines how to setup analyzer followed by an index, and can be cross references from the page. [not necessarily as part of scope of this PR]

------------

> ## Example

Author: @mgodwan Is this an appropriate example relevant to a developer using OpenSearch?

Reviewer: Yes, this looks good

For example, here's a sample configuration that defines a custom analyzer called `my_custom_analyzer`:

```json
PUT my_index
{
  "settings": {
    "analysis": {
      "analyzer": {
        "my_custom_analyzer": {
          "type": "custom",
          "tokenizer": "standard",
          "filter": [
            "lowercase",
            "my_stop_filter",
            "my_stemmer"
          ]
        }
      },
      "filter": {
        "my_stop_filter": {
          "type": "stop",
          "stopwords": ["the", "a", "and", "or"]
        },
        "my_stemmer": {
          "type": "stemmer",
          "language": "english"
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "my_text_field": {
        "type": "text",
        "analyzer": "my_custom_analyzer",
        "search_analyzer": "standard",
        "search_quote_analyzer": "my_custom_analyzer"
      }
    }
  }
}
```
{% include copy-curl.html %}

In this example, the `my_custom_analyzer` uses the standard tokenizer, converts all tokens to lowercase, applies a custom stopword filter, and then applies an English stemmer.

> You can then map a text field to use this custom analyzer for both indexing and searching:

Reviewer: Can we highlight the difference between analyzer, search_analyzer and search_quote_analyzer?

Author: Added text to existing documentation that explains them.

Author: Added note to direct user to existing source documentation.

---
layout: default

> title: analyzer

Author: ```suggestion
title: Analyzer
```

parent: Mapping parameters
grand_parent: Mapping and field types
nav_order: 5
has_children: false
has_toc: false
---

# `analyzer`

> The `analyzer` mapping parameter is used to define the text analysis process that applies to a text field during both indexing and searching operations.

Reviewer 2: "indexing and searching" => "index and search"?

The key functions of the `analyzer` mapping parameter are:

> 1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched. Each generated token must not exceed 32,766 bytes to avoid indexing failures.

Reviewer 2: ```suggestion
1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched. Each generated token must not exceed 32,766 bytes in order to avoid indexing failures.
```

2. **Normalization:** The analyzer can apply various normalization techniques, such as converting text to lowercase, removing stop words, and stemming/lemmatizing words.

> 3. **Consistency:** By defining the same analyzer for both indexing and searching, you ensure that the text analysis process is consistent, which helps improve the relevance of search results.

Reviewer 2: "indexing and searching" => "index and search operations"?

> 3. **Consistency:** By defining the same analyzer for both indexing and searching, you ensure that the text analysis process is consistent, which helps improve the relevance of search results.

Author: ```suggestion
3. **Consistency:** By defining the same analyzer for both index and search operations, you ensure that the text analysis process is consistent, which helps improve the relevance of search results.
```

> 4. **Customization:** OpenSearch allows you to define custom analyzers by specifying the tokenizer, character filters, and token filters to use. This gives you fine-grained control over the text analysis process.

Reviewer 2: ```suggestion
4. **Customization:** OpenSearch allows you to define custom analyzers by specifying the tokenizer, character filters, and token filters to be used. This gives you fine-grained control over the text analysis process.
```

> For information about specific analyzer parameters, such as `analyzer`, `search_analyzer`, and `search_quote_analyzer`, see [Search analyzers]({{site.url}}{{site.baseurl}}/analyzers/search-analyzers/).

Reviewer 2: ```suggestion
For information about specific analyzer parameters, such as `analyzer`, `search_analyzer`, or `search_quote_analyzer`, see [Search analyzers]({{site.url}}{{site.baseurl}}/analyzers/search-analyzers/).
```

{: .note}

------------

## Example

The following example configuration defines a custom analyzer called `my_custom_analyzer`:

```json
PUT my_index
{
  "settings": {
    "analysis": {
      "analyzer": {
        "my_custom_analyzer": {
          "type": "custom",
          "tokenizer": "standard",
          "filter": [
            "lowercase",
            "my_stop_filter",
            "my_stemmer"
          ]
        }
      },
      "filter": {
        "my_stop_filter": {
          "type": "stop",
          "stopwords": ["the", "a", "and", "or"]
        },
        "my_stemmer": {
          "type": "stemmer",
          "language": "english"
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "my_text_field": {
        "type": "text",
        "analyzer": "my_custom_analyzer",
        "search_analyzer": "standard",
        "search_quote_analyzer": "my_custom_analyzer"
      }
    }
  }
}
```
{% include copy-curl.html %}

In this example, the `my_custom_analyzer` uses the standard tokenizer, converts all tokens to lowercase, applies a custom stop word filter, and applies an English stemmer.

> You can then map a text field to use this custom analyzer for both indexing and searching:

Reviewer 2: "indexing and searching" => "index and search operations"? "to use" => "so that it uses"?

> You can then map a text field to use this custom analyzer for both indexing and searching:

Author: ```suggestion
You can then map a text field to use this custom analyzer for both index and search operations:
```

The key functions of the `analyzer` mapping parameter are:

1. **Tokenization:** The analyzer determines how the text is broken down into individual tokens (words, numbers) that can be indexed and searched. Each generated token must not exceed 32,766 bytes to avoid indexing failures.

2. **Normalization:** The analyzer can apply various normalization techniques, such as converting text to lowercase, removing stopwords, and stemming/lemmatizing words.

3. **Consistency:** By defining the same analyzer for both indexing and searching, you ensure that the text analysis process is consistent, which helps improve the relevance of search results.

4. **Customization:** OpenSearch allows you to define custom analyzers by specifying the tokenizer, character filters, and token filters to use. This gives you fine-grained control over the text analysis process.

For information about specific analyzer parameters, such as `analyzer`, `search_analyzer`, and `search_quote_analyzer`, see [Search analyzers]({{site.url}}{{site.baseurl}}/analyzers/search-analyzers/).
{: .note}

------------

## Example

> For example, here's a sample configuration that defines a custom analyzer called `my_custom_analyzer`:

Reviewer 3: ```suggestion
For example, the following sample configuration defines a custom analyzer called `my_custom_analyzer`:
```
